alf/networks/transformer_networks.py

```diff
@@ -0,0 +1,175 @@
+# Copyright (c) 2020 Horizon Robotics and ALF Contributors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import gin
+import torch
+import torch.nn as nn
+
+import alf
+from alf.networks import PreprocessorNetwork
+from alf.networks.memory import FIFOMemory
+from alf.nest.utils import NestConcat
+
+
+@gin.configurable
+class TransformerNetwork(PreprocessorNetwork):
+    """A Network composed of Memory and TransformerBlock.
+
+    The following is the pseudocode for the computation:
+
+    .. code-block:: python
+
+        for i in range(num_prememory_layers):
+            core, inputs = T_i([core, inputs], [core, inputs])
+        for j in range(num_memory_layers):
+            core, inputs = TM_j([memory_j, core, inputs], [core, inputs])
```

Comment:
This description is not accurate, because TransformerBlock only accepts tensors. Here it looks like lists.